bpf-map-pressure-exporter/src/collector.go

```diff
@@ -0,0 +1,40 @@
+package main
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type bpfMapPressureCollector struct {
+	describe *prometheus.Desc
+	fetcher  IBpfMapPressureFetcher
+}
+
+func newCollector(fetcher IBpfMapPressureFetcher) *bpfMapPressureCollector {
+	return &bpfMapPressureCollector{
+		describe: prometheus.NewDesc(
+			"bpf_map_pressure",
+			"bpf map pressure",
+			[]string{"map_id", "map_name"},
+			nil,
+		),
+		fetcher: fetcher,
+	}
+}
+
+func (c *bpfMapPressureCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.describe
+}
+
+func (c *bpfMapPressureCollector) Collect(ch chan<- prometheus.Metric) {
+	results := c.fetcher.Fetch()
+	for _, result := range results {
+		ch <- prometheus.MustNewConstMetric(
+			c.describe,
+			prometheus.GaugeValue,
+			result.mapPressure,
+			strconv.FormatUint(uint64(result.mapId), 10), result.mapName,
+		)
+	}
+}
```

Comment:
I think pressure calculation should be done asynchronously and the results are stored somewhere like a global variable, and `Collect()` only fetch map-pressure from that variable when it's called.
Doing calculation in the `Collect()` may cause a problem, because exporter's pressure is too high when the interval for metric collection is shortened and metrics collector gets timeout when the number of target bpf maps are so large.

Reply:
https://github.com/cybozu/neco-containers/pull/1155/commits/7f125beba4c58a82c05741e9bb42ad460474acac